### PR TITLE
Update integration & parallel test scripts to input the test type and reduce the number of input arguments

### PR DESCRIPTION
**Pull request by yantosca**
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard + GCST

### Confirm you have reviewed the following documentation

- [x] [Contributing guidelines](https://geos-chem.readthedocs.io/en/stable/help-and-reference/CONTRIBUTING.html)

### Describe the update
This PR follows upon PR #2182, which added the capability to run compile-only integration tests if the scheduler argument was set to `none`.  In this PR, we have made several additional updates:

1. Define a `testsToRun` bash variable in `test/integration/GCClassic/integrationTest.sh` and `test/integration/GCHP/integrationTest.sh`.  This is set to 'COMPILE` if called with `-t compile` and to `ALL` if called with `-t all`.

2. The `testsToRun` variable is passed as the 3rd argument to the corresponding `integrationTestCreate.sh` script.

3. Run directories for integration tests are not created when `testsToRun` is set to `NO`. 

4. The site (Harvard Cannon or Compute1) is now determined automatically by parsing the output of `uname -n`.

5. The environment file is only copied to the integration test directory structure and sourced if we are running integration tests on Harvard Cannon.

6. Functions `gcc_enable_or_disable_bootstrap` and `gchp_enable_or_disable_bootstrap` are not called if the `testToRun` variable is `COMPILE`.

7. TABs have been replaced with spaces in the `integrationTestCreate.sh` scripts for GCClassic and GCHP.

These updates will facilitate running compile-only tests as a GitHub action when pushes or pull requests are made to the superproject repos.  There will be corresponding PRs (yet to be made) in the GCHP and GCClassic superproject repos to add a YAML file to the `.github/workflows/` folder (which is needed to initialize the GitHub action).

### Expected changes

This is a zero-diff update.  Compile-only integration tests can now be submittted with, e.g.
```console
$ cd GCClassic/test/integration/GCClassic
$ ./integrationTest.sh -d /path/to/test/root/dir -t compile
```
which will result in the following output:
```console
==============================================================================
Creating GEOS-Chem Classic Integration Tests

GCClassic #42759a9 Add GitHub workflow for compile-only integration tests
GEOS-Chem #a5eaa3f3a Now create rundirs in int tests if local var $compileOnly is "no"
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release
==============================================================================
/n/home09/ryantosca/T/gc218x/GCC_componly is empty... nothing to clean up!

Creating CMake build directories:
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/default
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/apm
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/carbon
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/hg
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/luowd
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/rrtmg
 ... /n/home09/ryantosca/T/gc218x/GCC_componly/build/tomas15
Creating exe files directory /n/home09/ryantosca/T/gc218x/GCC_componly/bin
Creating env files directory /n/home09/ryantosca/T/gc218x/GCC_componly/env
Creating logs directory      /n/home09/ryantosca/T/gc218x/GCC_componly/logs
Creating scripts directory   /n/home09/ryantosca/T/gc218x/GCC_componly/scripts
Creating rundirs directory   /n/home09/ryantosca/T/gc218x/GCC_componly/rundirs
Linking to superproject      /n/home09/ryantosca/T/gc218x/GCC_componly/CodeDir

Copying run scripts to: /n/home09/ryantosca/T/gc218x/GCC_componly/scripts

Compiliation tests are running...
```

### Related Github Issue(s)

- See PR #2182

### CODE REVIEW

File: test/integration/GCClassic/integrationTestCreate.sh [hunk 1]
```diff
@@ -62,11 +65,11 @@ hemcoDir="${superProjectDir}/src/HEMCO"
...
 # Get the Git commit of the superproject and submodules
 head_gcc=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	   git -C "${superProjectDir}" log --oneline --no-decorate -1)
+           git -C "${superProjectDir}" log --oneline --no-decorate -1)
 head_gc=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	  git -C "${geosChemDir}" log --oneline --no-decorate -1)
+          git -C "${geosChemDir}" log --oneline --no-decorate -1)
 head_hco=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	   git -C "${hemcoDir}" log --oneline --no-decorate -1)
+           git -C "${hemcoDir}" log --oneline --no-decorate -1)
```
Comment (yantosca) [inline test/integration/GCClassic/integrationTestCreate.sh:72]: FYI, these 4 updates replaced TABs with spaces.

File: test/integration/GCHP/integrationTestCreate.sh [hunk 2]
```diff
@@ -62,11 +65,11 @@ hemcoDir="${superProjectDir}/src/GCHP_GridComp/HEMCO_GridComp/HEMCO"
...
 # Get the Git commit of the superproject and submodules
 head_gchp=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	    git -C "${superProjectDir}" log --oneline --no-decorate -1)
+            git -C "${superProjectDir}" log --oneline --no-decorate -1)
 head_gc=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	  git -C "${geosChemDir}" log --oneline --no-decorate -1)
+          git -C "${geosChemDir}" log --oneline --no-decorate -1)
 head_hco=$(export GIT_DISCOVERY_ACROSS_FILESYSTEM=1; \
-	   git -C "${hemcoDir}" log --oneline --no-decorate -1)
+           git -C "${hemcoDir}" log --oneline --no-decorate -1)
```
Comment (yantosca) [inline test/integration/GCHP/integrationTestCreate.sh:72]: These updates also replaced TABs with spaces.

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed:
```==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #42759a9 Add GitHub workflow for compile-only integration tests
GEOS-Chem #a5eaa3f3a Now create rundirs in int tests if local var $compileOnly is "no"
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

Using 24 OpenMP threads
Number of execution tests: 26

Submitted as SLURM job: 22839041
==============================================================================
 console


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Furthermore, all integration tests were zero-diff w/r/t 14.3.0, except for:
- APM (known parallelization issue)
- RRTMG (numerical noise in RRTMG collection file)

**Comment by yantosca** [comment]
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

GCHP      #2b21fe7 GEOS-Chem submod update: Merge PR #2182 (Add #SBATCH -c 1)
GEOS-Chem #a5eaa3f3a Now create rundirs in int tests if local var $compileOnly is "no"
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

Number of execution tests: 6

Submitted as SLURM job: 22839140
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Furthermore, all GCHP integration tests were zero-diff w/r/t 14.3.0

File: test/integration/GCClassic/README.md [hunk 3]
```diff
@@ -71,11 +71,9 @@ The integration test scripts accept the following command-line arguments
 
 `-e /path/to/env-file` specifies the file that is used to initialize the software environment.  This file will typically contain either `module load` or `spack load` commands.
 
-`-s scheduler` specifies the choice of scheduler (case-insensitive). SLURM and LSF schedulers are currently supported.  You may omit this for running interactively.
-
 `-p partition` specifies the partition for the scheduler.  You may omit this for running interactively.
 
-`-s` specifies that the SLURM scheduler will be used to run the integration test scripts.
+`-s scheduler` specifies that scheduler (`SLURM|LSF|none`) that will be used to run the integration test scripts.  Choosing `-s none` will run compilation-only tests.
```
Comment (lizziel) [inline test/integration/GCClassic/README.md:76]: "the scheduler" rather than "that scheduler"

**Review by lizziel — CHANGES_REQUESTED** [review]
It seems non-intuitive to control whether to run compile-only tests using the scheduler flag. Wouldn't it be more straightforward to pass a flag to configure this?

File: test/integration/GCClassic/README.md (hunk 3 above)
Comment (lizziel) [reply test/integration/GCClassic/README.md:76]: This comment applies to the GCHP README too.

**Comment by lizziel** [comment]
Adding on to my comment earlier, it seems like we want the following options:
- compile only and do not create test run dirs (does not require scheduler)
- compile and run (requires scheduler)

Based on this, what do you think of having a flag for test type (compile or all). If "compile", then scheduler would not be passed. If "all", then a valid scheduler must be passed and the program would exit with an error if not specified.

**Comment by yantosca** [comment]
> Based on this, what do you think of having a flag for test type (compile or all). If "compile", then scheduler would not be passed. If "all", then a valid scheduler must be passed and the program would exit with an error if not specified.

Yes, I like this approach and can implement it.

**Comment by yantosca** [comment]
@lizziel @msulprizio, I was thinking that we may be able to simplify this further by replacing `-s scheduler` with `-s site`, which would either be `cannon` or `compute1`.  Then the choice of site also dictates the partition and scheduler.  so then we would invoke the tests with

```console
./integration test.sh -t compile|all -s cannon|compute1  -e /path/to/env-file
```
For Cannon the partition would be `sapphire,huce_cascade,seas_compute,shared`.  We would have to change this (also with the run scripts in the `runScriptSamples/operational_examples/harvard_cannon` folder if things ever change.  But this doesn't happen that often so I think it's OK.

Also the `-e env-file` is only really needed for Cannon since on Compute1 everything is run in a container with the proper libraries already loaded.  So we can error-check this only if `-s cannon` is selected.

What do you think?

**Comment by lizziel** [comment]
I like this idea. For the case of Cannon we could even have a default environment path with the option to override it with -e. This would make typing out the test submission command even easier.

**Comment by yantosca** [comment]
Thanks @lizziel.  I'll work on this and push updates for your review.

**Comment by yantosca** [comment]
Also I just realized `site` is redundant.  We can test for it with bash function:
```bash
function get_site_name() {
    #========================================================================
    # Returns the site name based on the node name (i.e. uname -n)
    #========================================================================
    nodeName=$(uname -n)
    [[ "${nodeName}" =~ 'harvard.edu' ]] && echo "CANNON"
    [[ "${nodeName}" =~ 'wustl.edu'   ]] && echo "COMPUTE1"
    return 0
}
```

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #42759a9 Add GitHub workflow for compile-only integration tests
GEOS-Chem #053942b08 Bug fix: restore deleted $commonFuncs variable
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

Using 24 OpenMP threads
Number of execution tests: 26

Submitted as SLURM job: 23130876
==============================================================================
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also, all GEOS-Chem Classic integration tests were zero-diff w/r/t 14.3.0 except for:
- APM (known parallelization issue)
- RRTMG (numerical noise in RRTMG collection file)

**Comment by yantosca** [comment]
All GCHP integration tests also passed:
```console
==============================================================================
GCHP: Execution Test Results

GCHP      #2b21fe7 GEOS-Chem submod update: Merge PR #2182 (Add #SBATCH -c 1)
GEOS-Chem #864db3609 Fixed typo "gchp" -> "gcclassic" in GCHP integration test scripts
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

Number of execution tests: 6

Submitted as SLURM job: 23134086
==============================================================================


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also, all GCHP integration tests were zero-diff w/r/t 14.3.0.

File: test/integration/GCClassic/README.md (hunk 3 above)
Comment (yantosca) [reply test/integration/GCClassic/README.md]: @lizziel: This is now removed, as we no longer specify scheduler from the command line.

**Comment by yantosca** [comment]
I've made the similar updates in the parallel test scripts, which are now running.

**Comment by yantosca** [comment]
All parallelization tests passed:
```console
==============================================================================
GEOS-Chem Classic: Parallelization Test Results

GCClassic #42759a9 Add GitHub workflow for compile-only integration tests
GEOS-Chem #0f37d92f9 Update parallel test scripts for consistency w/ integration tests
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

1st run uses 24 OpenMP threads
2nd run uses 13 OpenMP threads
Number of parallelization tests: 4

Submitted as SLURM job: 23174479
==============================================================================
 
Parallelization tests:
------------------------------------------------------------------------------
gc_4x5_merra2_aerosol...............................Execute Simulation....PASS
gc_4x5_merra2_carbon................................Execute Simulation....PASS
gc_4x5_merra2_CH4...................................Execute Simulation....PASS
gc_4x5_merra2_fullchem..............................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Parallelization tests passed: 4
Parallelization tests failed: 0
Parallelization tests not yet completed: 0

==============================================================================
GEOS-Chem Classic: Parallelization Test Results

GCClassic #42759a9 Add GitHub workflow for compile-only integration tests
GEOS-Chem #0f37d92f9 Update parallel test scripts for consistency w/ integration tests
HEMCO     #f807e1a Update HEMCO version number to 3.8.0 in preparation for release

1st run uses 24 OpenMP threads
2nd run uses 13 OpenMP threads
Number of parallelization tests: 4

Submitted as SLURM job: 23174479
==============================================================================
 
Parallelization tests:
------------------------------------------------------------------------------
gc_4x5_merra2_aerosol...............................Execute Simulation....PASS
gc_4x5_merra2_carbon................................Execute Simulation....PASS
gc_4x5_merra2_CH4...................................Execute Simulation....PASS
gc_4x5_merra2_fullchem..............................Execute Simulation....PASS

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All parallelization tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
NOTE: This set of tests was done with `-q`, which only submits a short subset of all tests.

**Comment by yantosca** [comment]
NOTE: The minimum set of inputs is now:
```console
./integrationTest.sh -d /path/to/root/dir -t compile|all 
```

**Comment by yantosca** [comment]
Hi @msulprizio and @lizziel.  I've modified this PR so that we now copy utility scripts to the `utils/` folder of the integration test (or parallel test) root folder.  To restart a failed integration test you can now do:
```console
$ cd /path/to/integration/test/root
$ cd utils
$ ./redoIntegrationTestCompile.sh    # resubmits the compilation via SLURM
$ ./redoIntegrationTestExecute.sh    # resubmits the execution via SLURM
```
If you can give this PR another look, we can slip this into 14.3.1.  I have new integration & parallel tests running.

**Comment by msulprizio** [comment]
That it excellent. Thanks @yantosca. I will review today.

**Comment by yantosca** [comment]
BTW, the new integration tests for GCClassic and GCHP passed.

File: test/integration/GCHP/README.md [hunk 4]
```diff
@@ -17,76 +17,76 @@ Integration tests are short GCHP simulations.  Executables are compiled with all
...
+  - Driver script to run GEOS-Chem Classic integration tests
 - `integrationTestCreate.sh`
-  - Script to create GCHP run directories for integration tests
+  - Script to create GEOS-Chem Classic run directories for integration tests
 - `integrationTestCompile.sh`
-  - Script to compile GCHP executables
+  - Script to compile GEOS-Chem Classic executables
 - `integrationTestExecute.sh`
-   - Script to run GCHP integration test simulations
+   - Script to run GEOS-Chem Classic integration test simulations
```
Comment (msulprizio) [inline test/integration/GCHP/README.md:26]: These should be reverted to "GCHP" and not "GEOS-Chem Classic"

File: test/shared/commonFunctionsForTests.sh [hunk 5]
```diff
@@ -736,3 +737,45 @@ function gchp_enable_or_disable_bootstrap() {
...
+    echo "s/REQUESTED_PARTITION/${p}/"
+    return 0
+}
+
+
+function get_default_gcc_env_file() {
+    #========================================================================
+    # Returns the default environment file for GEOS-Chem Classic
+    #========================================================================
+    echo "/n/home09/ryantosca/repos/cannon-env/envs/gnu10/gcclassic.rocky+gnu10.minimal.env"
```
Comment (msulprizio) [inline test/shared/commonFunctionsForTests.sh:770]: I'm wondering if it would be better to put the env files in for example `run/GCClassic/runScriptSamples/operational_examples/harvard_cannon` rather than pointing to anyone's home directory. I notice there are env files for other clusters in that location.

**Review by msulprizio — CHANGES_REQUESTED** [review]

File: test/shared/commonFunctionsForTests.sh (hunk 5 above)
Comment (yantosca) [reply test/shared/commonFunctionsForTests.sh:770]: We can definitely do this.  I'll push.

Comment (yantosca) [reply test/shared/commonFunctionsForTests.sh]: @msulprizio: Now fixed in commits 920b062 and 2ea36ea

File: test/integration/GCHP/README.md (hunk 4 above)
Comment (yantosca) [reply test/integration/GCHP/README.md]: @msulprizio: Now fixed in 2ea36ea8425237606cc69011473af7277d9412b9

**Review by msulprizio — APPROVED** [review]

**Review by lizziel — APPROVED** [review]
Looks good!
